### PR TITLE
[BUILD] Re-enable ccache by default

### PR DESCRIPTION
**Pull request by tkonolige**
Previously ccache was disabled because of possible issues with hexagon. Re-enabling it to provide a best effort attempt at using it.

@driazati @cconvey

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -460,6 +460,42 @@ if(USE_PIPELINE_EXECUTOR)
   list(APPEND RUNTIME_SRCS ${RUNTIME_PIPELINE_SRCS})
 endif(USE_PIPELINE_EXECUTOR)
 
+#Caches the build.
+#Note that ccache-3.x doesn't support nvcc well, so CUDA kernels may never hit the cache and still
+#need to be re-compiled every time. Using ccache 4.0+ can resolve this issue.
+
+if(USE_CCACHE) # True for AUTO, ON, /path/to/ccache
```
Comment (driazati) [inline CMakeLists.txt:467]: this should be a `tvm_option` that defaults to `auto`

Comment (tkonolige) [reply CMakeLists.txt:467]: done

File: CMakeLists.txt [hunk 2]
```diff
@@ -460,6 +460,42 @@ if(USE_PIPELINE_EXECUTOR)
...
+        message(FATAL_ERROR "Cannot find ccache. Set USE_CCACHE mode to AUTO or OFF to build without ccache. USE_CCACHE=" "${USE_CCACHE}")
+      endif(CCACHE_FOUND)
+    else() # /path/to/ccache
+      set(PATH_TO_CCACHE USE_CCACHE)
+      message(STATUS "Setting ccache path to " "${PATH_TO_CCACHE}")
+    endif()
+    # Set the flag for ccache
+    if(DEFINED PATH_TO_CCACHE)
+      set(CXX_COMPILER_LAUNCHER PATH_TO_CCACHE)
+      set(C_COMPILER_LAUNCHER PATH_TO_CCACHE)
```
[the remark this answers is not in the document]
Comment (tkonolige) [reply CMakeLists.txt:494]: done

File: CMakeLists.txt [hunk 3]
```diff
@@ -460,6 +461,42 @@ if(USE_PIPELINE_EXECUTOR)
   list(APPEND RUNTIME_SRCS ${RUNTIME_PIPELINE_SRCS})
 endif(USE_PIPELINE_EXECUTOR)
 
+#Caches the build.
+#Note that ccache-3.x doesn't support nvcc well, so CUDA kernels may never hit the cache and still
+#need to be re-compiled every time. Using ccache 4.0+ can resolve this issue.
+
+if(USE_CCACHE) # True for AUTO, ON, /path/to/ccache
+  if(DEFINED CXX_COMPILER_LAUNCHER OR DEFINED C_COMPILER_LAUNCHER)
+    message(STATUS "CXX_COMPILER_LAUNCHER or C_COMPILER_LAUNCHER already defined, not using ccache")
```
Comment (areusch) [inline CMakeLists.txt:470]: should this be ERROR if USE_CCACHE is on?

File: CMakeLists.txt [hunk 4]
```diff
@@ -460,6 +461,42 @@ if(USE_PIPELINE_EXECUTOR)
...
+  else()
+    if("${USE_CCACHE}" STREQUAL "AUTO") # Auto mode
+      find_program(CCACHE_FOUND ccache)
+      if(CCACHE_FOUND)
+        message(STATUS "Found the path to ccache, enabling ccache")
+        set(PATH_TO_CCACHE ccache)
+      else()
+        message(STATUS "Didn't find the path to CCACHE, disabling ccache")
+      endif(CCACHE_FOUND)
+    elseif("${USE_CCACHE}" MATCHES ${IS_TRUE_PATTERN})
```
Comment (areusch) [inline CMakeLists.txt:480]: possible to collapse this with the previous if case?

**Review by areusch — COMMENTED** [review]
thanks @tkonolige

File: CMakeLists.txt (hunk 4 above)
Comment (tkonolige) [reply CMakeLists.txt:480]: I think the logic is clearer this way.

File: CMakeLists.txt [hunk 5]
```diff
@@ -460,6 +461,42 @@ if(USE_PIPELINE_EXECUTOR)
   list(APPEND RUNTIME_SRCS ${RUNTIME_PIPELINE_SRCS})
 endif(USE_PIPELINE_EXECUTOR)
 
+#Caches the build.
```
[the remark this answers is not in the document]
Comment (tkonolige) [reply CMakeLists.txt]: done

File: CMakeLists.txt [hunk 6]
```diff
@@ -460,6 +461,46 @@ if(USE_PIPELINE_EXECUTOR)
...
+    elseif("${USE_CCACHE}" MATCHES ${IS_TRUE_PATTERN})
+      find_program(CCACHE_FOUND ccache)
+      if(CCACHE_FOUND)
+        message(STATUS "Found the path to ccache, enabling ccache")
+        set(PATH_TO_CCACHE ccache)
+      else()
+        message(FATAL_ERROR "Cannot find ccache. Set USE_CCACHE mode to AUTO or OFF to build without ccache. USE_CCACHE=" "${USE_CCACHE}")
+      endif(CCACHE_FOUND)
+    else() # /path/to/ccache
+      set(PATH_TO_CCACHE USE_CCACHE)
```
Comment (driazati) [inline CMakeLists.txt:493]: same error with text substitution

File: CMakeLists.txt [hunk 7]
```diff
@@ -460,6 +461,46 @@ if(USE_PIPELINE_EXECUTOR)
   list(APPEND RUNTIME_SRCS ${RUNTIME_PIPELINE_SRCS})
 endif(USE_PIPELINE_EXECUTOR)
 
+# Caches the build.
```
Comment (driazati) [inline CMakeLists.txt:464]: organizational nit: can we move this into something like `cmake/ccache.cmake` and `include` it here instead?

File: CMakeLists.txt [hunk 8]
```diff
@@ -460,6 +461,46 @@ if(USE_PIPELINE_EXECUTOR)
...
+        message(FATAL_ERROR "Cannot find ccache. Set USE_CCACHE mode to AUTO or OFF to build without ccache. USE_CCACHE=" "${USE_CCACHE}")
+      endif(CCACHE_FOUND)
+    else() # /path/to/ccache
+      set(PATH_TO_CCACHE USE_CCACHE)
+      message(STATUS "Setting ccache path to " "${PATH_TO_CCACHE}")
+    endif()
+    # Set the flag for ccache
+    if(DEFINED PATH_TO_CCACHE)
+      set(CXX_COMPILER_LAUNCHER "${PATH_TO_CCACHE}")
+      set(C_COMPILER_LAUNCHER "${PATH_TO_CCACHE}")
```
Comment (driazati) [inline CMakeLists.txt:499]: can you add both of these vars to the header in `Summary.cmake` too?

**Review by driazati — APPROVED** [review]
a couple more nits but looks fine once those are addressed

**Comment by KOLANICH** [comment]
BTW, why not to use an external module?

**Comment by tkonolige** [comment]
@KOLANICH we don't want to add any unnecessary dependencies.

File: CMakeLists.txt (hunk 7 above)
Comment (tkonolige) [reply CMakeLists.txt:464]: done

File: CMakeLists.txt (hunk 8 above)
Comment (tkonolige) [reply CMakeLists.txt]: added

File: cmake/utils/CCache.cmake [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+if(USE_CCACHE) # True for AUTO, ON, /path/to/ccache
+  if(DEFINED CXX_COMPILER_LAUNCHER OR DEFINED C_COMPILER_LAUNCHER)
```
Comment (driazati) [inline cmake/utils/CCache.cmake:19]: also have no idea how cmake works but aren't these vars supposed to have the `CMAKE_` prefix everywhere?
```suggestion
  if(DEFINED CMAKE_CXX_COMPILER_LAUNCHER OR DEFINED CMAKE_C_COMPILER_LAUNCHER)
```

File: cmake/utils/Summary.cmake [hunk 10]
```diff
@@ -42,6 +42,7 @@ macro(print_summary)
     message(STATUS "  C++ compiler ID       : ${CMAKE_CXX_COMPILER_ID}")
     message(STATUS "  C++ compiler version  : ${CMAKE_CXX_COMPILER_VERSION}")
     message(STATUS "  CXX flags             : ${CMAKE_CXX_FLAGS}")
+    message(STATUS "  CXX launcher          : ${CMAKE_CXX_LAUNCHER}")
```
Comment (driazati) [inline cmake/utils/Summary.cmake:45]:
```suggestion
    message(STATUS "  CXX launcher          : ${CMAKE_CXX_COMPILER_LAUNCHER}")
```

File: cmake/utils/CCache.cmake (hunk 9 above)
Comment (tkonolige) [reply cmake/utils/CCache.cmake:19]: https://cmake.org/cmake/help/latest/prop_tgt/LANG_COMPILER_LAUNCHER.html. It appears to be CXX_COMPILER_LAUNCHER

**Review by areusch — APPROVED** [review]

**Pull request merged**